Empty variadic `tuple` subtype not assignable to `tuple[int, ...]`

```py
class Shape[*Ts](tuple[*Ts]): ...
class Shape0(Shape[()]): ...
class Shape1(Shape[int]): ...

def rank(shape: tuple[int, ...]) -> int:
    return len(shape)

_ = rank(Shape())       # ✅
_ = rank(())            # ✅
_ = rank(Shape0())      # ❌  Argument 1 to "rank" has incompatible type "Shape0"; expected "tuple[int, ...]"
_ = rank(Shape1([1]))   # ✅
```

[playground](https://mypy-play.net/?mypy=latest&python=3.13&flags=strict&gist=92dfe89c80f7e34d77d332b26cbfbd40)

Perhaps this is related to #19105, which also involves empty tuple subtypes.

I'm guessing that it's a consequence of this issue, but it might be worth noting that these variadic tuple subtypes ignore the constraints of typevars:

```pyi
class Shape[*Ts](tuple[*Ts]): ...

class Shape0(Shape[()]): ...
class Shape1(Shape[int]): ...
class ShapeN(Shape[*tuple[int, ...]]): ...


def upcast[T: (tuple[()], tuple[int], tuple[int, ...])](shape: T) -> T:
    return shape

reveal_type(upcast(Shape0()))  # ❌ Shape0
reveal_type(upcast(Shape1()))  # ✅ tuple[builtins.int]
reveal_type(upcast(ShapeN()))  # ❌ ShapeN
```

full output:

```
main.py:11: error: Value of type variable "T" of "upcast" cannot be "Shape0"  [type-var]
main.py:11: note: Revealed type is "__main__.Shape0"
main.py:12: note: Revealed type is "tuple[builtins.int]"
main.py:13: error: Value of type variable "T" of "upcast" cannot be "ShapeN"  [type-var]
main.py:13: note: Revealed type is "__main__.ShapeN"
Found 2 errors in 1 file (checked 1 source file)
```

[playground](https://mypy-play.net/?mypy=latest&python=3.13&flags=strict&gist=7e9f0d038409738afb182523f38f875f)

For reference, pyright reveals these as `tuple[()]`, `tuple[int]`, `tuple[int, ...]` ([playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=MYGwhgzhAEDKAWYAOBTA2gKgCoQLoAoAXAVyRHWzwEoAuaAOkYCgnRIYFkUAGfT1NPiq5aDZmyhxEqAIx9p6AJYA7QiLqN6rcJP4oAcvK6YSZJaoA0Y%2BrnXWWTACYoAZtFLBIhNFjpFS5ILCVqaBKmohAeaEVpoiBBAKdFhU0AC0AHzQvkzQedAATigkBcrQiVwsRQBuKGAgAPqEAJ6o%2BB5eRqi8VL15AMTQgKDk0KHoQrhMNXWNLW0dEIRdKHK9qdCDI2No4ZPT9U2tKO1InovLhmsDw6NRO5bWk0A)).

Let me know in case I should raise this separately.

--- a/mypy/maptype.py
+++ b/mypy/maptype.py
@@ -16,25 +16,6 @@ def map_instance_to_supertype(instance: Instance, superclass: TypeInfo) -> Insta
         # Fast path: `instance` already belongs to `superclass`.
         return instance
 
-    if superclass.fullname == "builtins.tuple" and instance.type.tuple_type:
-        if has_type_vars(instance.type.tuple_type):
-            # We special case mapping generic tuple types to tuple base, because for
-            # such tuples fallback can't be calculated before applying type arguments.
-            alias = instance.type.special_alias
-            assert alias is not None
-            if not alias._is_recursive:
-                # Unfortunately we can't support this for generic recursive tuples.
-                # If we skip this special casing we will fall back to tuple[Any, ...].
-                tuple_type = expand_type_by_instance(instance.type.tuple_type, instance)
-                if isinstance(tuple_type, TupleType):
-                    # Make the import here to avoid cyclic imports.
-                    import mypy.typeops
-
-                    return mypy.typeops.tuple_fallback(tuple_type)
-                elif isinstance(tuple_type, Instance):
-                    # This can happen after normalizing variadic tuples.
-                    return tuple_type
-
     if not superclass.type_vars:
         # Fast path: `superclass` has no type variables to map to.
         return Instance(superclass, [])
@@ -93,6 +74,28 @@ def map_instance_to_direct_supertypes(instance: Instance, supertype: TypeInfo) -
 
     for b in typ.bases:
         if b.type == supertype:
+
+            if supertype.fullname == "builtins.tuple" and instance.type.tuple_type:
+                if has_type_vars(instance.type.tuple_type):
+                    # We special case mapping generic tuple types to tuple base, because for
+                    # such tuples fallback can't be calculated before applying type arguments.
+                    alias = instance.type.special_alias
+                    assert alias is not None
+                    if not alias._is_recursive:
+                        # Unfortunately we can't support this for generic recursive tuples.
+                        # If we skip this special casing we will fall back to tuple[Any, ...].
+                        tuple_type = expand_type_by_instance(instance.type.tuple_type, instance)
+                        if isinstance(tuple_type, TupleType):
+                            # Make the import here to avoid cyclic imports.
+                            import mypy.typeops
+
+                            result.append(mypy.typeops.tuple_fallback(tuple_type))
+                            continue
+                        elif isinstance(tuple_type, Instance):
+                            # This can happen after normalizing variadic tuples.
+                            result.append(tuple_type)
+                            continue
+
             t = expand_type_by_instance(b, instance)
             assert isinstance(t, Instance)
             result.append(t)

--- a/mypy/typeanal.py
+++ b/mypy/typeanal.py
@@ -899,6 +899,7 @@ class TypeAnalyser(SyntheticTypeVisitor[Type], TypeAnalyzerPluginInterface):
                     ctx,
                     self.options,
                     use_standard_error=True,
+                    empty_tuple_index=empty_tuple_index,
                 )
             return tup.copy_modified(
                 items=self.anal_array(tup.items, allow_unpack=True), fallback=instance

--- a/test-data/unit/check-typevar-tuple.test
+++ b/test-data/unit/check-typevar-tuple.test
@@ -2803,3 +2803,38 @@ c2: S2[int, str] = S2((1, "", 2))  # E: Argument 1 to "S2" has incompatible type
 t: S2[Unpack[tuple[int, ...]]] = S2((1, 2))
 t_bad: S2[Unpack[tuple[int, ...]]] = S2((1, ""))  # E: Argument 1 to "S2" has incompatible type "tuple[int, str]"; expected "tuple[int, ...]"
 [builtins fixtures/tuple.pyi]
+
+[case testVariadicTupleSubclassEmptyIndex]
+from typing import TypeVarTuple, Unpack
+
+Ts = TypeVarTuple("Ts")
+class Shape(tuple[Unpack[Ts]]): ...
+
+class Shape0(Shape[()]): ...
+
+x: Shape[()]
+y: Shape0
+
+reveal_type(x)  # N: Revealed type is "tuple[(), fallback=__main__.Shape[()]]"
+reveal_type(y)  # N: Revealed type is "tuple[(), fallback=__main__.Shape0]"
+
+def test(t: tuple[int, ...]) -> None: ...
+test(x)
+test(y)
+[builtins fixtures/tuple.pyi]
+
+[case testVariadicTupleSubclassVariadicIndex]
+from typing import TypeVarTuple, Unpack
+
+Ts = TypeVarTuple("Ts")
+class Shape(tuple[Unpack[Ts]]): ...
+
+class ShapeN(Shape[Unpack[tuple[int, ...]]]): ...
+
+x: Shape[Unpack[tuple[int, ...]]]
+y: ShapeN
+
+def test(t: tuple[int, ...]) -> None: ...
+test(x)
+test(y)
+[builtins fixtures/tuple.pyi]

